packages/contracts/contracts/L1/AssetManager.sol

```diff
@@ -0,0 +1,1461 @@
+        emit KroUndelegationInitiated(validator, msg.sender, assets, shares);
+    }
+
+    /**
+     * @notice Initiate KGH undelegation for given validator and tokenId.
+     *
+     * @param validator Address of the validator.
+     * @param tokenId   Token id of KGH to undelegate.
+     */
+    function initUndelegateKgh(address validator, uint256 tokenId) external {
+        uint128 kghShares = _vault[validator].kghDelegators[msg.sender].kghShares[tokenId];
+        uint128 kroShares = _vault[validator].kghDelegators[msg.sender].kroShares[tokenId];
+
+        require(kghShares != 0, "AssetManager: No shares for the given tokenId");
+        _initUndelegateKgh(validator, msg.sender, tokenId, kroShares, kghShares);
+
+        emit KghUndelegationInitiated(validator, msg.sender, tokenId, kroShares, kghShares);
+    }
+
+    /**
+     * @notice Initiate KGH undelegation for given validator and tokenIds.
+     *
+     * @param validator Address of the validator.
+     * @param tokenIds  Array of token ids of KGHs to undelegate.
+     */
+    function initUndelegateKghBatch(address validator, uint256[] calldata tokenIds) external {
+        uint128 kroShares;
+        uint128 kghShares;
+        uint128 kroInKghs;
+        uint128 kghAssetsToWithdraw;
+        for (uint256 i = 0; i < tokenIds.length; ) {
+            uint128 kroSharesForTokenId = _vault[validator].kghDelegators[msg.sender].kroShares[
+                tokenIds[i]
+            ];
+            uint128 kghShareForTokenId = _vault[validator].kghDelegators[msg.sender].kghShares[
+                tokenIds[i]
+            ];
+
+            unchecked {
+                kroShares += kroSharesForTokenId;
+                kghShares += kghShareForTokenId;
+                kroInKghs += KGH_MANAGER.totalKroInKgh(tokenIds[i]);
+                kghAssetsToWithdraw += previewKghUndelegate(validator, tokenIds[i]);
+
+                delete _vault[validator].kghDelegators[msg.sender].kroShares[tokenIds[i]];
+                delete _vault[validator].kghDelegators[msg.sender].kghShares[tokenIds[i]];
+
+                ++i;
+            }
+        }
+
+        _initUndelegateKghBatch(
+            validator,
+            tokenIds,
+            kroInKghs,
+            kroShares,
+            kghShares,
+            kghAssetsToWithdraw
+        );
+
+        emit KghBatchUndelegationInitiated(validator, msg.sender, tokenIds, kroShares, kghShares);
+    }
+
+    /**
+     * @notice Finalize all pending KRO undelegation and returns the amount of assets that the vault would
+     *         exchange for the pending KRO shares.
+     *
+     * @param validator Address of the validator.
+     *
+     * @return The amount of assets that the vault would exchange for the pending KRO shares.
+     */
+    function finalizeUndelegate(address validator) public returns (uint128) {
+        if (_vault[validator].asset.totalPendingKroShares == 0) {
+            return 0;
+        }
```

Comment:
If there is no pending shares, why don't we revert this transaction?

Reply:
There can be an edge case since `finalizeUndelegateKgh` is using `finalizeUndelegate` inside itself. If a KGH delegator undelegates his KGH without receiving any reward, he cannot finalize his undelegation since his `pendingKroShare` is 0. 

But in the above conversation, I'm currently considering about managing the `pendingKroShares` separately for KGH undelegation, so if then it may be possible to add revert statement. Thanks for pointing this out!